persist/fs/commitlog/commit_log.go

```diff
@@ -0,0 +1,275 @@
+	closed         bool
+	closeErr       chan error
+}
+
+type commitLogWrite struct {
+	series       m3db.CommitLogSeries
+	datapoint    m3db.Datapoint
+	unit         xtime.Unit
+	annotation   []byte
+	attempt      int
+	completionFn completionFn
+}
+
+// NewCommitLog creates a new commit log
+func NewCommitLog(opts m3db.DatabaseOptions) (m3db.CommitLog, error) {
+	l := &commitLog{
+		opts:     opts,
+		nowFn:    opts.GetNowFn(),
+		writes:   make(chan commitLogWrite, commitLogWriteQueue),
+		metrics:  opts.GetMetricsScope().SubScope("commitlog"),
+		closeErr: make(chan error),
+	}
+
+	// Open the buffered commit log writer
+	if err := l.openWriter(l.nowFn()); err != nil {
+		return nil, err
+	}
+
+	// Flush the info header to ensure we can write to disk
+	if err := l.writer.Flush(); err != nil {
+		return nil, err
+	}
+
+	// Asynchronously write
+	go l.write()
+
+	return l, nil
+}
+
+func (l *commitLog) write() {
+	// TODO(r): also periodically flush the underlying commit log writer
+	// in case writes stall and commit log not written
+	var (
+		write commitLogWrite
+		retry = commitLogWrite{attempt: writeRetries}
+		open  bool
+	)
+	for {
+		if retry.attempt < writeRetries {
+			retry.attempt++
+			write = retry
+			time.Sleep(writeRetryDelay)
+		} else {
+			write, open = <-l.writes
+			if !open {
+				break
+			}
+		}
+
+		now := l.nowFn()
+		if !now.Before(l.writerExpireAt) {
+			if err := l.openWriter(now); err != nil {
+				l.metrics.IncCounter("writes.errors", 1)
+				l.metrics.IncCounter("writes.openerrors", 1)
+				// Re-attempt next write
+				retry = write
+				continue
+			}
+		}
+
+		// For writes requiring acks add to pending acks
+		if write.completionFn != nil {
+			l.pendingAcks = append(l.pendingAcks, write.completionFn)
+		}
+
+		err := l.writer.Write(write.series, write.datapoint, write.unit, write.annotation)
+		if err != nil {
+			l.metrics.IncCounter("writes.errors", 1)
+			// Log error and expire writer
+			l.writeError(err)
+			// Re-attempt next write
+			retry = write
+			continue
+		}
+
+		l.metrics.IncCounter("writes", 1)
+	}
+
+	if l.writer != nil {
+		writer := l.writer
+		l.writer = nil
+		l.closeErr <- writer.Close()
+	} else {
+		l.closeErr <- nil
+	}
+}
+
+func (l *commitLog) onFlush(err error) {
+	// Taking the pending acks is safe here as is called by the goroutine
+	// enqueuing the pending acks in write method from the writer as a
+	// result of a write
+	if len(l.pendingAcks) == 0 {
+		return
+	}
+
+	// Safe to callback acks as all they do is record a result and unblock
+	// another goroutine
+	for i := range l.pendingAcks {
+		l.pendingAcks[i](err)
+	}
+	l.pendingAcks = l.pendingAcks[:0]
+}
+
+func (l *commitLog) openWriter(now time.Time) error {
+	log := l.opts.GetLogger()
+	if l.writer != nil {
+		if err := l.writer.Close(); err != nil {
+			log.Errorf("failed to close commit log: %v", err)
+			// If we failed to close then create a new commit log writer
+			l.writer = nil
+		}
+	}
```

Comment:
It would be nice, definitely don't have the time to do that right now.  But yeah would be nice haha.